formfyxer/lit_explorer.py

```diff
@@ -50,6 +59,8 @@
 
     nlp = en_core_web_lg.load()
 
+passivepy = PassivePy.PassivePyAnalyzer("en_core_web_lg")
```

Comment:
This is super cool, but from [looking at the source a bit](https://github.com/mitramir55/PassivePy/blob/622062d265aa203ec4cc0de9a5e0eca30bc5f8dd/PassivePyCode/PassivePySrc/rules_for_all_passives.py#L9), I think it's loading an entirely new copy of the large spacy model that is just held in memory. Not sure how that would work with docassemble, but the spacy model is at least a few gigs, so simply importing anything from formfyxer would add an extra GB of memory usage, which would explain the dev server crashing with few users.

I think we can work this to use the same existing model that we already load, but we'd have to modify things upstream a bit. We can reach out over GitHub or directly to see if they'd like contributions, or if they'd want us to fork things instead.

Reply:
I can give it a try. I'm not sure of the best pattern though--I'm assuming we want to pass the `nlp` object as a parameter, so the same one can be used inside `formfyxer`?

We might also be able defer the initialization of `passivepy` so it's only started when it's needed, but then we might end up with it loading the model repeatedly.

Reply:
Passing the `nlp` object would be my first attempt. I do that if you defer the initialization, you should be able to do the global load pattern, i.e.:
```python
nlp = None
def my_func():
    global nlp
    if nlp is None:
        nlp = spacy.load(...)
    ...
```
And only load it once, but only if it's used. TBH I don't think that's enough though, because we'll still end up loading at least one `nlp` instance per user, which won't scale with any moderate use at all. It might need to be a longer running server somewhere else maybe? Not sure how we could set that up with docassemble

Reply:
Looking into this more, we could just make a flask endpoint that has a global `nlp` object loaded once, and hit it internally (would likely want to keep it internally). I'm not sure if that's just equivalent to docassemble global variables though. I do know that a copy would be made for each process, and Flask / uWSGI can spawn multiple processes (not sure how many docassemble spawns). Since it takes about 1.09 GB per instance, we only need a few (4ish) instances before we put a ton of memory pressure on the server (IIRC, we can peak around 4GB with just normal Weaver usage).

Reply:
Does this mean making a microservice that will process the text, essentially?

Reply:
Pretty much. We probably need to know a bit more about how DA handles global variables like that in interviews and how many processes might respond to the endpoint to know exactly how much it'd help. If neither of those turn out to help much, it'd mean a separate microservice.